docs/design/2023-06-16-runaway-queries-management.md

```diff
@@ -0,0 +1,110 @@
+Differentiating run-away queries from queries that really need to perform a full table/index scan is hard. There is no absolute rule. So we just let users define the rule to identify run-away queries. They can twist it on their own needs. The criteria are only the execution time, at least at present. Maybe add more dimension later.
+TiKV would send back the scan detail in coprocessor responses. If the total elapsed time of the query exceeds the threshold, then it would be recognized as a run-away query(statement).
+
+## User Interface
+
+### Create/Alter runaway rule
+
+```sql
+CREATE/ALTER RESOURCE GROUP rg1
+  RU_PER_SEC = 100000 [ PRIORITY = (HIGH|MEDIUM|LOW) ] [BURSTABLE]
+  [ QUERY_LIMIT = (
+         EXEC_ELAPSED = <#>,
+         ACTION = (DRYRUN|COOLDOWN|KILL),
+         [ WATCH = (EXACT|SIMILAR) DURATION <#> ])
+  ];
+```
+
+Runaway rules are stored in resource group meta.
+
+### Show runaway rule
+
+```sql
+mysql> SELECT * FROM information_schema.resource_groups;
++---------+------------+----------+-----------+-----------------------------------------------------------+                   
+| NAME    | RU_PER_SEC | PRIORITY | BURSTABLE |                       QUERY_LIMIT                         |      
++---------+------------+----------+-----------+-----------------------------------------------------------+                   
+| default | 1000000    | MEDIUM   | YES       | NULL                                                      |
+| rg1     | 2000       | HIGH     | YES       | EXEC_ELAPSED=10s, ACTION=KILL, WATCH=EXACT, DURATION=10m  |
+| rg2     | 1000       | HIGH     | YES       | EXEC_ELAPSED=10s, ACTION=KILL                             |
++---------+------------+----------+-----------+-----------------------------------------------------------+
+```
+
+### Log
+
+These fields are included for each record:  
+
+- TIME: sample time
+- RULE: "EXEC_ELAPSED"
+- ACTION: "DRYRUN"/"COOLDOWN"/"KILL"
+- MATCH: "RULE": identified by criteria, "WATCH": identified by matching queries
+- SQL_TEXT: text of the runaway sql
+- SQL_DIGEST: plan digest of the runaway sql
+
+#### Option1: persist in log file
+
+Runaway query records are stored in a dedicated, auto rotated, run-away log file, quite like slow log file. And the records can be also viewed by the admin table INFORMATION_SCHEMA.RUNAWAY_QUERIES which is a mapping to the run-away log file.
+
+#### Option2: persist in kv data (chosen)
+
+Print runaway query records in tidb log file, and also persist the records in KV data by the admin table `INFORMATION_SCHEMA.RUNAWAY_QUERIES`. The logs are persisted in batch with a flush interval config `runaway_queries_history_flush_interval`. Queries on the table `INFORMATION_SCHEMA.RUNAWAY_QUERIES` would return the results combining both in-memory and on-disk data.
```

Comment:
The current implementation is different from this description

Reply:
updated